vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,200 @@
+---
+title: "Data repository"
+author: "Thierry Onkelinx"
+date: "`r Sys.Date()`"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Data repository}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r setup, include = FALSE}
+library(knitr)
+opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+## Rationale
+
+In some cases we want to place data under version control. Data refers here to the information which is stored in a rectangular format. The columns define variables, while the rows represent observations. There is no information in neither the order of the columns nor the order of the rows. We assume that the data is available a an R `data.frame`.
+
+Although git can store binary files, its real power and efficiency is with plain text files. R has several functions to read and write `data.frames` as plain text files. Main downside is the loss of meta data, especially in case of `factors`. A `factor` will be written as a `character` and hence lose all information on the levels. This can be important when the levels are a) not sorted alphabetical, b) not all levels have observations or c) the `factor` is ordered. Besides that, storing a `factor` as `character` is not efficient.
+
+Git stores the version history under the form of diffs: a list of lines which are deleted and a list of lines which are inserted at a specific line number in a file. We have to keep this in mind when storing data as plain text files in Git. The table below indicate what impact changing the information content has on the diff in Git. Note that changing observations requires a small diff, while changing variables results in a massive diff.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("remove 1 observation", "remove 1 line"),
+    c("add 1 observation", "add 1 line"),
+    c("update 1 observation", "remove 1 line and add 1 line"),
+    c("remove 1 variable", "remove all lines and add all lines"),
+    c("add 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the information content of the data")
+```
```

Comment:
This is actually just a table to be written, so maybe use markdown table syntax itself? Now users could interpret this as a *R*-thing, whereas it is just a table:

|  Change in data | Git diff |
| ----------------------|----------|
| remove 1 observation | remove 1 line |
| add 1 observation | add 1 line |
| update 1 observation | remove 1 line and add 1 line |
| remove 1 variable | remove all lines and add all lines |
| add 1 variable | remove all lines and add all lines |

I do think this is more understandable as a table in the plain text version as well...

Reply:
I also prefer to use pandoc's markdown for plain tables, just to keep the process simple (except when the text table has to be retreived from somewhere). But no issue for me.